### ISSUE TITLE
[Media] new partials should work on edit view also

### ISSUE THREAD

**Issue by armababy** [AsgardCms/Platform#214]
### Summary of issue
Currently new way (event based) media is very cool, but ``HandleMediaStorage`` only attaches media to ``imageable``, so it won't update orders on old media and is not removing removed media :D

### Steps to reproduce
1) Create entity with ``@include('media::admin.fields.new-file-link-multiple', ['zone' => 'site_gallery'])``
2) Upload 2 or more media files
3) Edit created entity with ``@include('media::admin.fields.new-file-link-multiple', ['zone' => 'site_gallery'])``
4) Change order of media files and hit update
5) Order in database stays the same because of ``attach`` method

### System Details
- Operating System: Ubuntu 14.04
- PHP Version: 5.6.23
- Laravel Version: 5.1.40
- `Asgardcms/Platform` version: 1.14.1

**Comment by armababy** [comment]
I guess easy solution would be to use ``sync`` or new event handler for updating media.

**Comment by joshbrw** [comment]
Hm yeah.. It needs to be a "sync" really. This is a pretty common issue on
user interfaces where you can delete items, a lot of times we forget that
the relationship has to be detached so that the "removal" is done in the
database too.

**Comment by nWidart** [comment]
It's not an easy solution. there is more to it than just sync.

PR welcome if you want to make this happen. If not this is not a priority.

**Comment by armababy** [comment]
Okay, i'll work on this i have similar stuff already on some of my stuff.

**Comment by armababy** [comment]
Hey, take a look [at this branch](https://github.com/armababy/Media/tree/feature/prepare-new-partials)
Tell me if i'm forgetting anything, if you can test run it.

Basically when this is done we could remove old file partials infavour of new ones

Changes/Details:
* Added method ``filesByZone`` in ``Modules\Media\Support\Traits\MediaRelation``
* Allows to have many single and multiple new partial includes (independant processing for each zone)
* File is unique for each zone (can't have two identical files in same zone) - later one will overwrite previous ones
* Every zone can have same file
* Load initial order to preserve ordering on pivot sync
* Some typo fixes

Outstanding issues:
* Removing media does not update orders input
* Adding new media does not update orders input

Improvements needed:
* Handle order updates in javascript on all media actions (remove, add, sort, submit form) - when this is done we can remove initial order state calculations from multiple view

**Comment by nWidart** [comment]
Hello,

Thank you for your work!

- Can be used on create, edit views using 1 event (as described in docs)
    - I think you mean the event handler right ? Since you used `sync` instead of attach in the event handler class. 

    It looks good though.
- Helper method `filesByZone`, very nice!
- Allows to have many single and multiple new partial includes (independant processing for each zone)
    - That's the `5dcea7fb` commit I assume ? From what I can see it looks good. 
- The jquery UI fix: 👍 

About the needed improvements, it would indeed be nice to have all actions edit the order of the files.

**Comment by armababy** [comment]
Will not rush with any PR yet, want to triple check and spot any problems this time :D
Would be nice to sort out dragging problems with multiple media partials and also re-ordering of orders.
Yeah basically because it now processes medias by zones instead of one-by-one we can have any number of zones (single and/or multiple), i'm testing with 2/3 single and 2/3 multi new media partials.

**Comment by armababy** [comment]
Another problem to consider:
When including many media fields, the styles and all of the javascript gets inserted many times, that's probably not so cool to trash dom like that.

Possible solution use style like 
```
@include('media::admin.fields.fresh-file-link-multiple', ['zones' => ['site_gallery', 'site_media']])
```
This way we can return styles and javascript only once. But that's only suitable if user  wants media fields one after another.

Another possible solution would be to include styles and javascript beforehand and remove them from media views.
```
@include('media::admin.fields.helpers.single')
@include('media::admin.fields.helpers.multiple')

@include('media::admin.fields.fresh-file-link-single', ['zone' => 'site_logo'])
<!-- Some html here -->
@include('media::admin.fields.fresh-file-link-multiple', ['zone' => 'site_gallery'])
<!-- Some html here -->
@include('media::admin.fields.fresh-file-link-single', ['zone' => 'site_image'])
```
But this again makes things much more complicated and confusing for user.

Anyone got idea on this one?
Main target is to allow for user to use as many single and/or multiple media field includes where styles and javascript is inluded only once and it's not too confusing to explain to people.

**Comment by armababy** [comment]
Added update orders(zone specific) on insert, sort, remove actions.
I'm not good with javascript or jquery so if anyone proficient with it I'm happy to optimize.
[Commit in question](https://github.com/armababy/Media/commit/6a44b0c1fd37648035e6a9ea9d47f823f144ae6d)

**Comment by nWidart** [comment]
Haha you and me both :P @motchju is a pro-javascript, he did most of the JS part here.

**Comment by armababy** [comment]
Theres one [interesting solving](https://github.com/armababy/Media/commit/4667a20541ce2425acb0651fdf793706e0d3ccce) of the multiple include styles and scripts as i described in comment before.

I do not yet know for sure how you like tour Providers to be served, so i made new one didn't bloat Media one.

I think this is most painful way to do this from Asgard users perspective, nothing changes for them.

**Comment by armababy** [comment]
Hmm sounds cool, remove completely all scripts and styles from include view and use dedicated asset files. Would have to extract things like ``{!! route('media.grid.select') !!}`` as it will be pure js file.

Does Asgard have some sort of global object where you could put stuff like that?
Or i could leavel 1 line of javascript variable with grid select route

**Comment by nWidart** [comment]
There is indeed a global object https://github.com/AsgardCms/AdminLTE/blob/2.0/views/partials/asgard-globals.blade.php Depending on when it's needed in the app life cycle it could be added there, or in the view as a one liner indeed

**Comment by armababy** [comment]
Yeah i will see the differences and then either make several js files or combine them into one.

**Comment by nWidart** [comment]
The end goal was also to use only the 'event' way of linking media, meaning also on the edit partials. 
I just didn't have time currently, plus the current edit views work so wasn't a priority.

Even better would be to have 1 single partial and 1 multiple partial, for both create and edit.

**Comment by armababy** [comment]
I'll make my thing work for now :) Then we can look into eventing all of this.
I just want any number of media partials included new event way for create and edit with no problems etc.

So i successfully extracted styles and javascripts to their own assets and included them via asset manager and pipeline.

I can merge everything to single files (media-include.css, media-include.js or something like that), but as i understand you want 2 of each for future customization (media-single-include.css, media-multiple-include.css, media-single-include.js, media-multiple-include.js)?

**Comment by armababy** [comment]
Something like that i guess then: [Asset resolution with assetManager/Pipe](https://github.com/armababy/Media/commit/f5f654286d889c645bf5a82573b33c53ce59c724)
For that to work you would need to publish assets (which happens anyway on every update run)

Also would require adding this ``mediaGridSelect: '{{ route('media.grid.select') }}'`` to Asgard javascript global 


I did not do this for old media partials (file-link.blade, file-link-multiple.blade) because of all those zones and entityClass etc, can't really extract at stage it's currently on.
But that does not matter for now, both work, but aim is to deprecate or straight up remove old style of adding media partials. 
Since new event one can handle any amount of media includes, pipelining assets, calculating order on all actions and sync with database.
